### ISSUE TITLE
OpenID Connect: compare cannocialized versions of redirect_uri

### ISSUE THREAD

**Issue by AugierLe42e** [diaspora/diaspora#6643]
Registering a client with url `https://example.org/` will cause implicit flow to throw `invalid client id or redirect uri` if `https://example.org` is used.

**Comment by jhass** [comment]
Could you rephrase? This is very confusing.

**Comment by AugierLe42e** [comment]
Registering a client with `https://example.org/` as a redirect URI will make the token requests to fail if you use `https://example.org` instead of `https://example.org/` in further requests. Nothing is specified about trailing slashes in the spec. As `https://example.org/` and `https://example.org` are always treated as the same URI, it's a bit counter-intuitive to raise an exception when `https://example.org` is used instead of `https://example.org/`.

**Comment by jhass** [comment]
Mmmh, I see, yes we should compare the canonicalized URI

**Comment by theworldbright** [comment]
Hmm the [spec](http://openid.net/specs/openid-connect-core-1_0.html#AuthRequest) seems to say the follows:

> redirect_uri
> REQUIRED. Redirection URI to which the response will be sent. This URI MUST exactly match one of the Redirection URI values for the Client pre-registered at the OpenID Provider, with the matching performed as described in Section 6.2.1 of [RFC3986] (Simple String Comparison).

where the link takes you to

> 6.2.1.  Simple String Comparison
> 
>    If two URIs, when considered as character strings, are identical,
>    then it is safe to conclude that they are equivalent.

which seems to be different from

> 6.2.3.  Scheme-Based Normalization
> 
>    The syntax and semantics of URIs vary from scheme to scheme, as
>    described by the defining specification for each scheme.
>    Implementations may use scheme-specific rules, at further processing
>    cost, to reduce the probability of false negatives.  For example,
>    because the "http" scheme makes use of an authority component, has a
>    default port of "80", and defines an empty path to be equivalent to
>    "/", the following four URIs are equivalent:
> 
>       http://example.com
>       http://example.com/
>       http://example.com:/
>       http://example.com:80/
> 
>    In general, a URI that uses the generic syntax for authority with an
>    empty path should be normalized to a path of "/".  Likewise, an
>    explicit ":port", for which the port is empty or the default for the
>    scheme, is equivalent to one where the port and its ":" delimiter are
>    elided and thus should be removed by scheme-based normalization.  For
>    example, the second URI above is the normal form for the "http"
>    scheme.
> 

Although personally, I don't see why we shouldn't use the scheme-based one.

**Comment by AugierLe42e** [comment]
To me, anything that reduces frictions in usage is a good thing to me. We all know that sometimes web developers can be careless but the problem will fall on the final user. So, IMHO, let's follow the  Scheme-Based Normalization. Notice that the section 6.2.1 says that two URI's are equivalent if they are identical when compared as char strings. It doesn't say that they aren't if they are different when compared as char strings. So section 6.2.1 doesn't contredict section 6.2.3.

**Comment by CSammy** [comment]
Just an idea:
> [...] This URI MUST exactly match one of the Redirection URI values for the Client pre-registered at the OpenID Provider [...]

So a way according to spec would be to register both URI values, with and without trailing slash. This should be possible as per [Client Registration Spec](http://openid.net/specs/openid-connect-registration-1_0.html#ClientMetadata).

I don't believe it's valid to argue that a comparison by means of 6.2.3 would not violate the spec because 6.2.1 does not explicitly exclude it. The spec refers explicitly to comparison by means of 6.2.1, so that is the spec way. If you do it differently, it's not according to spec.

Whether that is a problem or not is a different matter, but the claim it'd be still according to spec is not valid.

**Comment by jhass** [comment]
Doing proper normalization seems to add quite a lot of complexity and it's actually not an issue the end user will face but the app developer should already face in their testing. Since as @CSammy says the spec is quite explicitly referring to 6.2.1, this is a won't fix.
